db/index.go

```diff
@@ -0,0 +1,495 @@
+
+	txn, err := c.getTxn(ctx, false)
+	if err != nil {
+		return err
+	}
+	_, err = c.getIndexes(ctx, txn)
+	if err != nil {
+		return err
+	}
+	for i := range c.indexes {
+		if c.indexes[i].Name() == indexName {
+			err = c.indexes[i].RemoveAll(ctx, txn)
+			if err != nil {
+				return err
+			}
+			c.indexes = append(c.indexes[:i], c.indexes[i+1:]...)
+			break
+		}
+	}
+	err = txn.Systemstore().Delete(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *collection) dropAllIndexes(ctx context.Context) error {
+	prefix := core.NewCollectionIndexKey(c.Name(), "")
+	txn, err := c.getTxn(ctx, false)
+	if err != nil {
+		return err
+	}
+	q, err := txn.Systemstore().Query(ctx, query.Query{
+		Prefix: prefix.ToString(),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := q.Close(); err != nil {
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	for res := range q.Next() {
+		if res.Error != nil {
+			return res.Error
+		}
+		err = txn.Systemstore().Delete(ctx, ds.NewKey(res.Key))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func deserializePrefix[T any](ctx context.Context, prefix string, storage ds.Read) ([]T, error) {
+	q, err := storage.Query(ctx, query.Query{Prefix: prefix})
+	if err != nil {
+		return nil, NewErrFailedToCreateCollectionQuery(err)
+	}
+	defer func() {
+		if err := q.Close(); err != nil {
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	elements := make([]T, 0)
+	for res := range q.Next() {
+		if res.Error != nil {
+			return nil, res.Error
+		}
+
+		var element T
+		err = json.Unmarshal(res.Value, &element)
+		if err != nil {
+			return nil, NewErrInvalidStoredIndex(err)
+		}
+		elements = append(elements, element)
+	}
+	return elements, nil
+}
+
+func (c *collection) getIndexes(ctx context.Context, txn datastore.Txn) ([]CollectionIndex, error) {
+	if c.isIndexCached {
+		return c.indexes, nil
+	}
+
+	prefix := core.NewCollectionIndexKey(c.Name(), "")
+	if txn == nil {
+		var err error
+		txn, err = c.getTxn(ctx, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+	indexes, err := deserializePrefix[client.IndexDescription](ctx, prefix.ToString(), txn.Systemstore())
+	if err != nil {
+		return nil, err
+	}
+	colIndexes := make([]CollectionIndex, 0, len(indexes))
+	for _, index := range indexes {
+		colIndexes = append(colIndexes, NewCollectionIndex(c, index))
+	}
+
+	c.indexes = colIndexes
+	c.isIndexCached = true
+	return colIndexes, nil
+}
+
+func (c *collection) GetIndexes(ctx context.Context) ([]client.IndexDescription, error) {
+	indexes, err := c.getIndexes(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	indexDescriptions := make([]client.IndexDescription, 0, len(indexes))
+	for _, index := range indexes {
+		indexDescriptions = append(indexDescriptions, index.Description())
+	}
+
+	return indexDescriptions, nil
+}
+
+func (c *collection) createIndex(
```

Comment:
I know you moved this to `collection_index.go` but the same comment applies. It would be nice to have `createIndex(...)` under the `CreateIndex(...)` function.